pallets/msa/src/lib.rs

```diff
@@ -883,6 +976,90 @@ impl<T: Config> Pallet<T> {
 		}
 		Ok(Some(schemas.0.into()))
 	}
+
+	/// Add a signature to the PayloadSignatureRegistry based on a virtual "bucket" grouping.
+	/// First reset the virtual bucket if necessary, then add the new entry.
+	pub fn register_signature(
+		signature: &MultiSignature,
+		current_block: T::BlockNumber,
+		mortality_block: T::BlockNumber,
+	) -> DispatchResult {
+		if Self::mortality_block_limit(current_block) < mortality_block {
+			Err(Error::<T>::MortalityTooHigh.into())
+		} else {
+			let bucket_num = Self::bucket_for(mortality_block.into());
+			<PayloadSignatureRegistry<T>>::try_mutate(
+				bucket_num,
+				signature,
+				|maybe_mortality_block| -> DispatchResult {
+					ensure!(maybe_mortality_block.is_none(), Error::<T>::BucketKeyExists);
+
+					Self::reset_virtual_bucket_if_needed(mortality_block, bucket_num)?;
+					*maybe_mortality_block = Some(mortality_block);
+					Ok(())
+				},
+			)
+		}
+	}
+
+	// Check if enough blocks have passed to reset bucket mortality storage.
+	// If so:
+	//     1. delete all the stored bucket/signature alues with key1 = bucket num
+	//     2. set the bucket's mortality block to the new value
+	// If not, don't do anything.
+	fn reset_virtual_bucket_if_needed(
+		mortality_block: T::BlockNumber,
+		bucket_num: T::BlockNumber,
+	) -> DispatchResult {
+		<MortalityBlockOf<T>>::try_mutate(bucket_num, |maybe_mortality_block| -> DispatchResult {
+			let bucket_mortality_block = Self::bucket_mortality_block(mortality_block);
+
+			match maybe_mortality_block {
+				None => Err(Error::<T>::NoSuchBucket.into()),
+				Some(mortality_block_for_bucket) => {
+					if mortality_block > *mortality_block_for_bucket {
+						Self::delete_signatures_for_bucket(bucket_num)?;
+						*maybe_mortality_block = Some(bucket_mortality_block);
+					} else if *mortality_block_for_bucket == T::BlockNumber::default() {
+						*maybe_mortality_block = Some(bucket_mortality_block);
+					}
+					Ok(())
+				},
+			}
+		})
+	}
+
+	// delete signatures with the given bucket number since they have existed past the mortality
+	// block limit.
+	fn delete_signatures_for_bucket(bucket_num: T::BlockNumber) -> Result<(), DispatchError> {
+		// use drain_prefix
+		// https://paritytech.github.io/substrate/master/frame_support/pallet_prelude/struct.StorageDoubleMap.html#method.drain_prefix
+		// we dont' need the value; it's just to finalize the drain.
+		<PayloadSignatureRegistry<T>>::drain_prefix(bucket_num).count();
```

Comment:
I'm not sure how `drain_prefix` is implemented but if it uses the underlying logic as `remove_prefix` it should be fine, if not maybe switching with `remove_prefix` would be more efficient  https://substrate.recipes/double.html

Reply:
thanks I'll look at that some more.

Reply:
Looks like `remove_prefix` is deprecated. They tell you to use `clear_prefix` instead, which I had overlooked, so I'll use that.